File: desktop-mend/src/test/java/commands/SetupTest.java
package commands;

import co.samco.mend4.core.AppProperties;
import co.samco.mend4.core.Settings;
import co.samco.mend4.core.bean.EncodedKeyInfo;
import co.samco.mend4.desktop.commands.Setup;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.invocation.InvocationOnMock;
import org.mockito.stubbing.Answer;

import javax.crypto.BadPaddingException;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.KeyPair;
import java.security.NoSuchAlgorithmException;
import java.security.spec.InvalidKeySpecException;
import java.util.Arrays;
import java.util.Collections;

import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyString;
import static org.mockito.Mockito.*;

public class SetupTest extends TestBase {
    private Setup setup;

    private ArgumentCaptor<String> errCaptor;

    private final String mendDirPath = "settingspath";
    private final String settingsFilePath = mendDirPath + "/settings.file";

    @Before
    public void setup() {
        super.setup();
        errCaptor = ArgumentCaptor.forClass(String.class);
        when(fileResolveHelper.getSettingsFilePath()).thenReturn(settingsFilePath);
        when(fileResolveHelper.getMendDirPath()).thenReturn(mendDirPath);
        setup = new Setup(log, strings, osDao, keyHelper, cryptoProvider, fileResolveHelper, settings);
    }

    @Test
    public void alreadySetUp() {
        when(osDao.fileExists(any(File.class))).thenReturn(true);
        setup.execute(Collections.emptyList());
        verify(err).println(errCaptor.capture());
        Assert.assertEquals(strings.getf("SetupMend.alreadySetup", settingsFilePath, Setup.FORCE_FLAG),
                errCaptor.getValue());
    }

    @Test
    public void wrongArgNum3() {
        setup.execute(Arrays.asList("a", "b", "c"));
        verify(err).println(errCaptor.capture());
        Assert.assertEquals(strings.getf("General.invalidArgNum", Setup.COMMAND_NAME), errCaptor.getValue());
    }

    @Test
    public void wrongArgNum1() {
        setup.execute(Arrays.asList("a"));
        verify(err).println(errCaptor.capture());
        Assert.assertEquals(strings.getf("General.invalidArgNum", Setup.COMMAND_NAME), errCaptor.getValue());
    }

    @Test
    public void passwordsDontMatch() throws NoSuchAlgorithmException, NoSuchPaddingException, InvalidKeyException,
            IOException, IllegalBlockSizeException, BadPaddingException, InvalidAlgorithmParameterException,
            InvalidKeySpecException  {
        String passOne = "passOne";
        String passTwo = "passTwo";
        when(osDao.readPassword(anyString())).thenAnswer(new Answer<char[]>() {
            int count = 0;
            @Override
            public char[] answer(InvocationOnMock invocation) {
                if (count++ == 0) {
                    return passOne.toCharArray();
                } else {
                    return passTwo.toCharArray();
                }
            }
        });
        EncodedKeyInfo keyInfo = new EncodedKeyInfo("a", "b", 0);
        when(cryptoProvider.getEncodedKeyInfo(any(char[].class), any(KeyPair.class))).thenReturn(keyInfo);
        setup.execute(Collections.emptyList());
        verify(err, times(2)).println(errCaptor.capture());
        verifySettingsSetup(keyInfo);
        Assert.assertEquals(strings.get("SetupMend.passwordMismatch"), errCaptor.getAllValues().get(0));
        Assert.assertEquals(strings.get("SetupMend.complete"), errCaptor.getAllValues().get(1));
    }

    @Test
    public void setupFromKeyFiles() throws NoSuchPaddingException, IOException,
            InvalidKeyException, NoSuchAlgorithmException, IllegalBlockSizeException, BadPaddingException,
            InvalidAlgorithmParameterException, InvalidKeySpecException {
        EncodedKeyInfo keyInfo = doSetupFromKeyFiles();
        verifySettingsSetup(keyInfo);
        Assert.assertEquals(strings.get("SetupMend.complete"), errCaptor.getValue());
    }

    @Test
    public void setupFromKeyFilesNull () throws FileNotFoundException {
        String exception = "exception";
        when(osDao.readPassword(anyString())).thenReturn("password".toCharArray());
        when(fileResolveHelper.resolveFile(anyString())).thenThrow(new FileNotFoundException(exception));
        setup.execute(Arrays.asList("x", "y"));
        verify(err).println(errCaptor.capture());
        Assert.assertEquals(exception, errCaptor.getValue());
    }

    @Test
    public void setupSettingsThrowsException () throws NoSuchPaddingException, IOException, InvalidKeyException,
            NoSuchAlgorithmException, IllegalBlockSizeException, BadPaddingException, InvalidAlgorithmParameterException,
            InvalidKeySpecException {
        String exception = "exception";
        when(osDao.readPassword(anyString())).thenReturn("password".toCharArray());
        EncodedKeyInfo keyInfo = new EncodedKeyInfo("a", "b", 0);
        when(cryptoProvider.getEncodedKeyInfo(any(char[].class), any(KeyPair.class))).thenReturn(keyInfo);
        doThrow(new IOException(exception)).when(settings).setValue(any(Settings.Name.class), anyString());
        setup.execute(Arrays.asList("x", "y"));
        verify(err).println(errCaptor.capture());
        Assert.assertEquals(exception, errCaptor.getValue());
    }

    @Test
    public void doesntOverrideSettingsAlreadySet() throws NoSuchPaddingException, InvalidAlgorithmParameterException,
            NoSuchAlgorithmException, IOException, BadPaddingException, IllegalBlockSizeException, InvalidKeyException,
            InvalidKeySpecException {
        when(settings.valueSet(eq(Settings.Name.PREFERREDAES))).thenReturn(true);
        doSetupFromKeyFiles();
        verify(settings, never()).setValue(eq(Settings.Name.PREFERREDAES), anyString());
        Assert.assertEquals(strings.get("SetupMend.complete"), errCaptor.getValue());
    }

    public EncodedKeyInfo doSetupFromKeyFiles() throws NoSuchPaddingException, IOException,
            InvalidKeyException, NoSuchAlgorithmException, IllegalBlockSizeException, BadPaddingException,
            InvalidAlgorithmParameterException, InvalidKeySpecException {
        when(osDao.readPassword(anyString())).thenReturn("password".toCharArray());
        EncodedKeyInfo keyInfo = new EncodedKeyInfo("a", "b", 0);
        when(cryptoProvider.getEncodedKeyInfo(any(char[].class), any(KeyPair.class))).thenReturn(keyInfo);
        when(fileResolveHelper.resolveFile(anyString())).thenReturn(new File(""));
        setup.execute(Arrays.asList("x", "y"));
        verify(keyHelper).readKeyPairFromFiles(any(File.class), any(File.class));
        verify(err).println(errCaptor.capture());
        return keyInfo;
    }

    private void verifySettingsSetup(EncodedKeyInfo keyInfo) throws IOException {
        verify(settings).setValue(eq(Settings.Name.PREFERREDAES), eq(AppProperties.PREFERRED_AES_ALG));
        verify(settings).setValue(eq(Settings.Name.PREFERREDRSA), eq(AppProperties.PREFERRED_RSA_ALG));
        verify(settings).setValue(eq(Settings.Name.AESKEYSIZE), eq(Integer.toString(AppProperties.PREFERRED_AES_KEY_SIZE)));
        verify(settings).setValue(eq(Settings.Name.RSAKEYSIZE), eq(Integer.toString(keyInfo.getKeySize())));
        verify(settings).setValue(eq(Settings.Name.PRIVATEKEY), eq(keyInfo.getPrivateKey()));
        verify(settings).setValue(eq(Settings.Name.PUBLICKEY), eq(keyInfo.getPublicKey()));
    }

}

File: desktop-mend/src/test/java/commands/SetupTest.kt
package commands

import co.samco.mend4.core.Settings
import co.samco.mend4.desktop.commands.Setup
import co.samco.mend4.desktop.dao.SettingsDao
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.mockito.ArgumentMatchers.anyString
import com.nhaarman.mockitokotlin2.KArgumentCaptor
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.argumentCaptor
import com.nhaarman.mockitokotlin2.atLeastOnce
import com.nhaarman.mockitokotlin2.eq
import com.nhaarman.mockitokotlin2.never
import com.nhaarman.mockitokotlin2.times
import com.nhaarman.mockitokotlin2.verify
import com.nhaarman.mockitokotlin2.whenever
import java.io.File
import java.io.IOException
import java.security.KeyPair

class SetupTest : TestBase() {

    private lateinit var setup: Setup
    private lateinit var errCaptor: KArgumentCaptor<String>
    private lateinit var outCaptor: KArgumentCaptor<String>
    private val mendDirPath = File("settingspath")
    private val settingsFilePath = File(mendDirPath, "settings.file")

    @Before
    override fun setup() {
        super.setup()
        errCaptor = argumentCaptor()
        outCaptor = argumentCaptor()
        whenever(fileResolveHelper.settingsFile).thenReturn(settingsFilePath)
        whenever(fileResolveHelper.mendDirFile).thenReturn(mendDirPath)

        setup = Setup(
            log = log,
            strings = strings,
            cryptoProvider = cryptoProvider,
            fileResolveHelper = fileResolveHelper,
            settings = settings,
            osDao = osDao
        )
    }

    @Test
    fun alreadySetUp() {
        whenever(osDao.exists(any())).thenReturn(true)
        setup.execute(emptyList())
        verify(err).println(errCaptor.capture())
        Assert.assertEquals(
            strings.getf("SetupMend.alreadySetup", settingsFilePath, Setup.FORCE_FLAG),
            errCaptor.firstValue
        )
    }

    @Test
    fun wrongArgNum3() {
        setup.execute(listOf("a", "b", "c"))
        verify(err).println(errCaptor.capture())
        Assert.assertEquals(
            strings.getf("General.invalidArgNum", Setup.COMMAND_NAME),
            errCaptor.firstValue
        )
    }

    @Test
    fun wrongArgNum1() {
        setup.execute(listOf("a"))
        verify(err).println(errCaptor.capture())
        Assert.assertEquals(
            strings.getf("General.invalidArgNum", Setup.COMMAND_NAME),
            errCaptor.firstValue
        )
    }

    @Test
    fun passwordsDontMatch() {
        val passOne = "passOne"
        val passTwo = "passTwo"
        whenever(osDao.readLine()).thenReturn(" ")
        whenever(osDao.exists(eq(settingsFilePath))).thenReturn(false)
        var count = 0
        whenever(osDao.readPassword(anyString())).thenAnswer {
            return@thenAnswer if (count++ == 0) {
                passOne.toCharArray()
            } else {
                passTwo.toCharArray()
            }
        }

        setup.execute(emptyList())

        verify(err, times(1)).println(errCaptor.capture())
        verify(cryptoProvider, never()).storeEncryptedKeys(any(), any())
        Assert.assertEquals(strings["SetupMend.passwordMismatch"], errCaptor.allValues[0])
    }

    @Test
    fun setupDefaultParameters() {
        doSetup()
        verifySettingsSetup()
        Assert.assertEquals(strings["SetupMend.complete"], outCaptor.lastValue)
    }

    @Test
    fun setupSettingsThrowsException() {
        val exception = "exception"
        whenever(osDao.readLine()).thenReturn(" ")
        whenever(osDao.readPassword(anyString()))
            .thenReturn("password".toCharArray())
        var settingsExists = false
        whenever(settings.setValue(any(), anyString())).thenAnswer {
            settingsExists = true
            throw IOException(exception)
        }
        whenever(osDao.exists(eq(settingsFilePath)))
            .thenAnswer { settingsExists }

        setup.execute(emptyList())

        verify(err).println(errCaptor.capture())
        verify(osDao).delete(eq(settingsFilePath))
        Assert.assertEquals(exception, errCaptor.firstValue)
    }

    private fun doSetup() {
        whenever(osDao.readPassword(anyString())).thenReturn("password".toCharArray())
        whenever(osDao.readLine()).thenReturn(" ")
        whenever(cryptoProvider.generateKeyPair())
            .thenReturn(KeyPair(null, null))

        setup.execute(emptyList())

        verify(cryptoProvider).generateKeyPair()
        verify(cryptoProvider, times(1))
            .storeEncryptedKeys(eq("password".toCharArray()), any())
        verify(out, atLeastOnce()).println(outCaptor.capture())
    }

    private fun verifySettingsSetup(
        asymmetricCipherName: String = Setup.DEFAULT_ASYMMETRIC_CIPHER,
        asymmetricCipherTransform: String = Setup.DEFAULT_ASYMMETRIC_CIPHER_TRANSFORM,
        asymmetricKeySize: Int = Setup.DEFAULT_KEY_SIZE,
        pwKeyFactoryIterations: Int = Setup.DEFAULT_PW_FACTORY_ITERATIONS,
        pwKeyFactoryParallelism: Int = Setup.DEFAULT_PW_FACTORY_PARALLELISM,
        pwKeyFactoryMemory: Int = Setup.DEFAULT_PW_FACTORY_MEMORY_KB,
    ) {
        verify(settings).setValue(
            eq(Settings.Name.ASYMMETRIC_CIPHER_NAME),
            eq(asymmetricCipherName)
        )
        verify(settings).setValue(
            eq(Settings.Name.ASYMMETRIC_CIPHER_TRANSFORM),
            eq(asymmetricCipherTransform)
        )
        verify(settings).setValue(
            eq(Settings.Name.ASYMMETRIC_KEY_SIZE),
            eq(asymmetricKeySize.toString())
        )
        verify(settings).setValue(
            eq(Settings.Name.PW_KEY_FACTORY_ITERATIONS),
            eq(pwKeyFactoryIterations.toString())
        )
        verify(settings).setValue(
            eq(Settings.Name.PW_KEY_FACTORY_PARALLELISM),
            eq(pwKeyFactoryParallelism.toString())
        )
        verify(settings).setValue(
            eq(Settings.Name.PW_KEY_FACTORY_MEMORY_KB),
            eq(pwKeyFactoryMemory.toString())
        )

        val logDir = File(mendDirPath, Setup.DEFAULT_LOG_DIR_NAME)
        val encDir = File(mendDirPath, Setup.DEFAULT_ENC_DIR_NAME)
        val decDir = File(mendDirPath, Setup.DEFAULT_DEC_DIR_NAME)
        val shredCommand: String = Setup.DEFAULT_SHRED_COMMAND
        val currentLog: String = Setup.DEFAULT_LOG_FILE_NAME

        verify(osDao).mkdirs(eq(mendDirPath))
        verify(osDao).mkdirs(eq(logDir))
        verify(osDao).mkdirs(eq(encDir))
        verify(osDao).mkdirs(eq(decDir))

        verify(settings).setValue(
            eq(SettingsDao.LOG_DIR),
            eq(logDir.absolutePath)
        )
        verify(settings).setValue(
            eq(SettingsDao.ENC_DIR),
            eq(encDir.absolutePath)
        )
        verify(settings).setValue(
            eq(SettingsDao.DEC_DIR),
            eq(decDir.absolutePath)
        )
        verify(settings).setValue(
            eq(SettingsDao.SHRED_COMMAND),
            eq(shredCommand)
        )
        verify(settings).setValue(
            eq(SettingsDao.CURRENT_LOG),
            eq(currentLog)
        )
    }
}